bridge/gitlab/export.go

```diff
@@ -1,27 +1,504 @@
+	ge.identityClient = make(map[string]*gitlab.Client)
+	ge.cachedOperationIDs = make(map[string]string)
+
 	return nil
 }
 
+// getIdentityClient return a gitlab v4 API client configured with the access token of the given identity.
+// if no client were found it will initialize it from the known tokens map and cache it for next use
+func (ge *gitlabExporter) getIdentityClient(id entity.Id) (*gitlab.Client, error) {
+	client, ok := ge.identityClient[id.String()]
+	if ok {
+		return client, nil
+	}
+
+	// get token
+	token, ok := ge.identityToken[id.String()]
+	if !ok {
+		return nil, ErrMissingIdentityToken
+	}
+
+	// create client
+	client = buildClient(token)
+	// cache client
+	ge.identityClient[id.String()] = client
+
+	return client, nil
+}
+
 // ExportAll export all event made by the current user to Gitlab
-func (ge *gitlabExporter) ExportAll(repo *cache.RepoCache, since time.Time) (<-chan core.ExportResult, error) {
-	return nil, nil
+func (ge *gitlabExporter) ExportAll(ctx context.Context, repo *cache.RepoCache, since time.Time) (<-chan core.ExportResult, error) {
+	out := make(chan core.ExportResult)
+
+	user, err := repo.GetUserIdentity()
+	if err != nil {
+		return nil, err
+	}
+
+	ge.identityToken[user.Id().String()] = ge.conf[keyToken]
+
+	// get repository node id
+	ge.repositoryID = ge.conf[keyProjectID]
+
+	go func() {
+		defer close(out)
+
+		var allIdentitiesIds []entity.Id
+		for id := range ge.identityToken {
+			allIdentitiesIds = append(allIdentitiesIds, entity.Id(id))
+		}
+
+		allBugsIds := repo.AllBugsIds()
+
+		for _, id := range allBugsIds {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				b, err := repo.ResolveBug(id)
+				if err != nil {
+					out <- core.NewExportError(err, id)
+					return
+				}
+
+				snapshot := b.Snapshot()
+
+				// ignore issues created before since date
+				// TODO: compare the Lamport time instead of using the unix time
+				if snapshot.CreatedAt.Before(since) {
+					out <- core.NewExportNothing(b.Id(), "bug created before the since date")
+					continue
+				}
+
+				if snapshot.HasAnyActor(allIdentitiesIds...) {
+					// try to export the bug and it associated events
+					ge.exportBug(ctx, b, since, out)
+				} else {
+					out <- core.NewExportNothing(id, "not an actor")
+				}
+			}
+		}
+	}()
+
+	return out, nil
+}
+
+// exportBug publish bugs and related events
+func (ge *gitlabExporter) exportBug(ctx context.Context, b *cache.BugCache, since time.Time, out chan<- core.ExportResult) {
+	snapshot := b.Snapshot()
+
+	var err error
+	var bugGitlabID int
+	var bugGitlabIDString string
+	var bugCreationId string
+
+	// Special case:
+	// if a user try to export a bug that is not already exported to Gitlab (or imported
+	// from Gitlab) and we do not have the token of the bug author, there is nothing we can do.
+
+	// skip bug if origin is not allowed
+	origin, ok := snapshot.GetCreateMetadata(core.KeyOrigin)
+	if ok && origin != target {
```

Comment:
Maybe instead of adding `gitlab-url` metadata we can directly add `gitlab-project-id`. Since using `gitlab-url` will force us to request the gitlab API to get the project ID

Reply:
whatever you choose, it must be the most immutable identifier possible. You can (should) also store more than one even if you only use one.

Reply:
done